src/coreclr/src/vm/domainfile.h

```diff
@@ -755,7 +753,100 @@ class DomainAssembly : public DomainFile
 
 typedef DomainAssembly::ModuleIterator DomainModuleIterator;
 
-// --------------------------------------------------------------------------------
-// DomainModule is a subclass of DomainFile which specifically represents a module.
-// --------------------------------------------------------------------------------
+// ------------------------------------------------------------------------------------------------------
+// DomainCompositeImage is a subclass of DomainFile which specifically represents a composite R2R module.
+// ------------------------------------------------------------------------------------------------------
+
+struct AssemblyNameIndex
+{
+    SString Name;
+    int32_t Index;
+    
+    AssemblyNameIndex() : Index(-1) {}
+    AssemblyNameIndex(const SString& name, int32_t index);
+};
+
+class AssemblyNameIndexHashTraits : public NoRemoveSHashTraits< PtrSHashWithCleanupTraits<AssemblyNameIndex, const SString&> >
+{
+public:
+    static AssemblyNameIndex *Null() { return NULL; }
+    static bool IsNull(const AssemblyNameIndex *e) { return e == NULL; }
+
+    static const SString& GetKey(const AssemblyNameIndex *assemblyNameIndex) { return assemblyNameIndex->Name; }
+    static BOOL Equals(const SString& a, const SString& b) { return a.Equals(b); }
+    static count_t Hash(const SString& a) { return a.Hash(); }
+};
+
+class DomainCompositeImage : public DomainFile
```

Comment:
I am not sure whether inheriting `DomainCompositeImage` from `DomainFile` is the best design. It makes it have number of concepts that do not make sense for it. `DomainFile::GetExposedModuleObject`, `DomainFile::IsSystem`, the load levels, ...  none of it is applicable to the composite native image.

Would it be better if this is `class NativeImage` that does not inherit from anything?

Reply:
I have no problems with that whatsoever. I just had to start somewhere so I picked a place that seemed doing something remotely similar :-)… made the more interesting by the fact that JanV whom I'd normally consult on such matters first was OOF this week. Will move and refactor, sure.

Reply:
Refactored in 7th commit.